src/Symfony/Component/Messenger/Transport/MemoryTransport.php

```diff
@@ -0,0 +1,64 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Transport;
+
+use Symfony\Component\Messenger\Asynchronous\Transport\ReceivedMessage;
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\MessageBusInterface;
+
+class MemoryTransport implements TransportInterface
+{
+    private $envelopes = array();
+    private $bus;
+
+    public function __construct(MessageBusInterface $bus)
+    {
+        $this->bus = $bus;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function receive(callable $handler): void
+    {
+        foreach ($this->envelopes as $envelope) {
```

Comment:
Can you `while(count($this->envelopes) > 0)` and `array_pop` in here? That way you are sure the envelopes have been consumed (i.e. nothing remaining in `$this->envelopes` at the end and only the non-processed ones in case of exception).

Reply:
`array_pop` or `array_shift`? Shouldn't they be processed in the order they were sent?

Reply:
Good shot, shift sounds good 👍